app/change_set_persisters/change_set_persister/extract_archival_collection_code.rb

```diff
@@ -0,0 +1,34 @@
+# frozen_string_literal: true
+class ChangeSetPersister
+  class ExtractArchivalCollectionCode
+    attr_reader :change_set_persister, :change_set
+    def initialize(change_set_persister:, change_set:, post_save_resource: nil)
+      @change_set = change_set
+      @change_set_persister = change_set_persister
+    end
+
+    def run
+      return unless change_set.respond_to?(:source_metadata_identifier)
+      return unless change_set.model.respond_to?(:archival_collection_code)
+      return unless updated_values
+      return if PulMetadataServices::Client.bibdata?(change_set.source_metadata_identifier)
+      change_set.model.archival_collection_code = extract_collection_code(change_set.source_metadata_identifier)
+      change_set
+    end
+
+    private
+
+      # either the source metadata identifier has changed or we're refreshing remote metadata
+      def updated_values
+        change_set.changed?(:source_metadata_identifier) || (
+          change_set.respond_to?(:apply_remote_metadata?) &&
+          change_set.apply_remote_metadata?
+        )
+      end
+
+      def extract_collection_code(pulfa_id)
+        m = pulfa_id.match(/^([a-zA-Z]+[0-9]+)_.*/)
```

Comment:
I don't think this regex will match all of the possible EAD variants, which include:
* `RCPXG-5830371.2`
* `C0744.04`
* `C0723.1-47`
* `C0723.306e`
* `MC001.03.03`
* `MC197.r16845`

Maybe something like `/^([A-Za-z0-9.-]+)_c.*/` would work for everything?

Reply:
okay thanks for catching this. I added a test to show all those cases.